### PR TITLE
feat: support partition prune api

### PR DESCRIPTION
**Pull request by KnightChess**
## Description

Add filtering capabilities to table API, currently only partition fields are applicable. Multiple predicates are `AND` together.

```rust
hudi_table.read_snapshot(&["foo != a", "bar >= 100"]);
```
Supported operators are: `>, >=, <, <=, =, !=`.

For #47 

## How are the changes test-covered

- [ ] N/A
- [x] Automated tests (unit and/or integration tests)
- [ ] Manual tests
- [ ] Details are described below

### CODE REVIEW

**Comment by KnightChess** [comment]
@xushiyan cc

**Comment by KnightChess** [comment]
@xushiyan hello, Is there a problem with my test method in my local?
image
image

**Comment by KnightChess** [comment]
> hello, Is there a problem with my test method in my local?
ignore checkstyle, fetch the latest commit, it work again, but can not report python test error.

**Comment by KnightChess** [comment]
@xushiyan cc:

**Comment by xushiyan** [comment]
@KnightChess awesome contribution! let me take a look. i might push some quick fixes just FYI to move faster.

File: crates/core/Cargo.toml [hunk 1]
```diff
@@ -63,8 +63,9 @@ tokio = { workspace = true }
 # datafusion
 datafusion = { workspace = true, optional = true }
 datafusion-expr = { workspace = true, optional = true }
-datafusion-common = { workspace = true, optional = true }
+datafusion-common = { workspace = true }
 datafusion-physical-expr = { workspace = true, optional = true }
+log = "0.4.22"
```
Comment (xushiyan) [inline crates/core/Cargo.toml:68]: can this dep be added to workspace for centralized management?

File: crates/core/Cargo.toml [hunk 2]
```diff
@@ -63,8 +63,9 @@ tokio = { workspace = true }
 # datafusion
 datafusion = { workspace = true, optional = true }
 datafusion-expr = { workspace = true, optional = true }
-datafusion-common = { workspace = true, optional = true }
+datafusion-common = { workspace = true }
```
Comment (xushiyan) [inline crates/core/Cargo.toml:66]: hudi-core should be optionally dependent on datafusion - only when datafusion feature is enabled. let's keep this optional

File: crates/core/src/table/partitions.rs [hunk 3]
```diff
@@ -0,0 +1,362 @@
...
+use std::collections::HashMap;
+
+/// A special value used in Hive to represent the null partition in partitioned tables
+pub const NULL_PARTITION_VALUE_DATA_PATH: &str = "__HIVE_DEFAULT_PARTITION__";
+
+/// A Enum used for selecting the partition value operation when filtering a HudiTable partition.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum PartitionValue {
+    /// The partition value with the equal operator
+    Equal(ScalarValue),
```
Comment (xushiyan) [inline crates/core/src/table/partitions.rs:34]: echoing comment at the beginning - can we remove the dependency of ScalarValue from datafusion? wouldn't we be ok to keep strings in the model and convert using DataType for comparison when matching partitions? use something from arrow if suitable

File: crates/core/src/table/fs_view.rs [hunk 4]
```diff
@@ -70,7 +73,31 @@ impl FileSystemView {
...
+        if partition_filters.is_empty() {
+            return Ok(partition_paths);
+        }
+        let field_and_data_type: HashMap<_, _> = partition_schema
+            .fields()
+            .iter()
+            .map(|field| (field.name().to_string(), field.data_type().clone()))
+            .collect();
+
+        Ok(partition_paths
```
Comment (xushiyan) [inline crates/core/src/table/fs_view.rs:85]: all partition paths were listed and then filtered - a more optimized approach is to filter out partitions during the listing process, will reduce a lot of scanning for multi-level partitions. we can optimize this in a separate PR

**Review by xushiyan — CHANGES_REQUESTED** [review]
@KnightChess good progress. i think the main functionalities are there. A major change is to re-model the Partition struct, we need to decouple it from datafusion struct

File: crates/core/src/table/mod.rs [hunk 5]
```diff
@@ -234,6 +239,23 @@ impl Table {
...
 
+    pub async fn get_partition_schema(&self) -> Result<Schema> {
+        let schema = self.get_schema().await;
+        let partition_str_vec = self
+            .configs
+            .get_or_default(PartitionFields)
+            .to::<Vec<String>>();
+        let partition_set: HashSet<String> = partition_str_vec.into_iter().collect();
+        let partition_fields: Vec<Arc<Field>> = schema
+            .unwrap()
```
[the remark this answers is not in the document]
Comment (xushiyan) [reply crates/core/src/table/mod.rs]: this was fixed.

File: crates/core/src/table/mod.rs [hunk 6]
```diff
@@ -234,6 +239,23 @@ impl Table {
         self.timeline.get_latest_schema().await
     }
 
+    pub async fn get_partition_schema(&self) -> Result<Schema> {
+        let schema = self.get_schema().await;
+        let partition_str_vec = self
```
[the remark this answers is not in the document]
Comment (xushiyan) [reply crates/core/src/table/mod.rs]: this was fixed.

File: crates/core/src/table/mod.rs [hunk 7]
```diff
@@ -306,6 +330,16 @@ impl Table {
...
+
+    pub fn partition_filter_replace(&mut self, partition_filters: Vec<PartitionFilter>) {
+        if self.configs.get_or_default(IsHiveStylePartitioning).to() {
+            self.partition_filters = partition_filters;
+        }
+    }
+
+    pub fn reset(&mut self) {
+        self.file_system_view.reset()
+    }
```
[the remark this answers is not in the document]
Comment (xushiyan) [reply crates/core/src/table/mod.rs:345]: @KnightChess in this PR we can annotate them as test helper like `#[cfg(test)]`, then you can file a separate PR to refactor the test so we won't allow mutability through these table API.

**Comment by xushiyan** [comment]
@KnightChess do you think you can address the main comment in the next few days? then i can polish further if needed and land this. Trying to get this in the upcoming release within 2 weeks 🙂

**Comment by KnightChess** [comment]
@xushiyan sorry for reply late, I will address these two days

File: crates/core/src/table/mod.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (KnightChess) [reply crates/core/src/table/mod.rs:345]: @xushiyan This is not only for testing, but also to refresh the cache, like spark need use `refresth table defualt.testA` to refresth cache meta.

File: crates/core/Cargo.toml (hunk 1 above)
Comment (KnightChess) [reply crates/core/Cargo.toml:68]: oh, forgot it, will update it.

File: crates/core/Cargo.toml (hunk 2 above)
Comment (KnightChess) [reply crates/core/Cargo.toml:66]: It now relies on the `ScalarValue` of DataFusion.

File: crates/core/src/table/partitions.rs (hunk 3 above)
Comment (KnightChess) [reply crates/core/src/table/partitions.rs:34]: thanks, I will try to remove `the dependency of ScalarValue` and fix `datafusion-common = { workspace = true }` be optional.

File: crates/core/src/table/fs_view.rs (hunk 4 above)
Comment (KnightChess) [reply crates/core/src/table/fs_view.rs:85]: good idea, will create a separate PR improve it.

File: crates/core/src/table/mod.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (xushiyan) [reply crates/core/src/table/mod.rs:345]: @KnightChess yes understood, eventually we'll need to implement those functionalities. But let's stick to the on-demand principle - don't "predict" the needs, adding APIs when it becomes necessary. As of the codebase, these 2 serve for testing so let's write proper test code, and we can evolve them for production code whenever need to.

**Comment by KnightChess** [comment]
@xushiyan Hello, I couldn't find an implementation similar to ScalarValue, and I am not very familiar with Arrow yet, and. There is a certain learning curve involved, which might delay the progress of this PR. Could you please help improve this PR?

**Comment by KnightChess** [comment]
@xushiyan cc, I try to use arrow `Scalar<ArrayRef>` to replace datafusion `ScalarValue`, and modified some suggestions to repair.

File: crates/core/src/config/table.rs [hunk 8]
```diff
@@ -130,6 +130,7 @@ impl ConfigParser for HudiTableConfig {
         match self {
             Self::DatabaseName => Some(HudiConfigValue::String("default".to_string())),
             Self::DropsPartitionFields => Some(HudiConfigValue::Boolean(false)),
+            Self::PartitionFields => Some(HudiConfigValue::List(vec!["".to_string()])),
```
Comment (xushiyan) [inline crates/core/src/config/table.rs:133]: default as a list of an empty str ? should it be an empty list

File: crates/core/src/table/partitions.rs [hunk 9]
```diff
@@ -289,59 +267,6 @@ impl TryFrom<(&str, &HashMap<String, DataType>)> for HudiTablePartition {
...
 mod tests {
     use super::*;
-    use serde_json::json;
-
-    fn check_json_serialize(filter: PartitionFilter, expected_json: &str) {
-        assert_eq!(serde_json::to_value(filter).unwrap(), json!(expected_json))
-    }
-
-    #[test]
-    fn test_serialize_partition_filter() {
```
Comment (xushiyan) [inline crates/core/src/table/partitions.rs]: @KnightChess why the serialization test was removed?

Comment (KnightChess) [reply crates/core/src/table/partitions.rs]: @xushiyan arrow scalar not implement `Display`, and I found the scenes that are not used at present, so I remove `Serialize` trait. If necessary, I can add implementation

File: crates/core/Cargo.toml [hunk 10]
```diff
@@ -73,6 +74,5 @@ hudi-tests = { path = "../tests" }
 datafusion = [
     "dep:datafusion",
     "datafusion-expr",
-    "datafusion-common",
```
[the remark this answers is not in the document]
Comment (KnightChess) [reply crates/core/Cargo.toml:76]: I forgot to add it back. It seems that the test has a issue, and there is no error without this dependence.

File: crates/core/src/table/partitions.rs (hunk 9 above)
Comment (KnightChess) [reply crates/core/src/table/partitions.rs]: @xushiyan thank you so much

**Comment by xushiyan** [comment]
I was wrapping up my vacation 😄  just now getting back to update this:

Changes I've made:
- Add `PartitionPruner` to handle pruning logic, taking care of hudi configs like hive style and url encoded paths.
- Taking filters from parameters in form of `foo > 10`, `bar != a`, etc, only apply `AND` to them.
- Limit the operators to only `>, >=, <, <=, =, !=` to avoid rabbit hole of implementing sql parser; `IN` and `NOT IN` can be achieved anyway.
- API to reset should not be needed now, as the pruning with different filters should be applied to loaded fs view directly.

There are more follow up work to do on datafusion integration side, which I'll jot down in the GH issue.

**Review by xushiyan — APPROVED** [review]

**Comment by KnightChess** [comment]
@xushiyan thanks review

**Pull request merged**